src/middlewared/middlewared/plugins/nfs.py

```diff
@@ -61,3 +64,233 @@ def nfs_compress(self, nfs):
         self.nfs_extend(new)
 
         return new
+
+
+class SharingNFSService(CRUDService):
+    class Config:
+        namespace = "sharing.nfs"
+        datastore = "sharing.nfs_share"
+        datastore_prefix = "nfs_"
+        datastore_extend = "sharing.nfs.extend"
+
+    @accepts(Dict(
+        "sharingnfs_create",
+        List("paths", items=[Dir("path")]),
+        Str("comment"),
+        List("networks", items=[IPAddr("network", cidr=True)]),
+        List("hosts", items=[IPAddr("host")]),
+        Bool("alldirs"),
+        Bool("ro"),
+        Bool("quiet"),
+        Str("maproot_user", required=False),
+        Str("maproot_group", required=False),
+        Str("mapall_user", required=False),
+        Str("mapall_group", required=False),
+        List("security", items=[Str("provider", enum=["sys", "krb5", "krb5i", "krb5p"])]),
+        register=True,
+    ))
+    async def do_create(self, data):
+        verrors = ValidationErrors()
+
+        await self.validate(data, "sharingnfs_create", verrors)
+
+        if verrors:
+            raise verrors
+
+        await self.compress(data)
+        paths = data.pop("paths")
+        data["id"] = await self.middleware.call(
+            "datastore.insert", self._config.datastore, data,
+            {
+                "prefix": self._config.datastore_prefix
+            },
+        )
+        for path in paths:
+            await self.middleware.call(
+                "datastore.insert", "sharing.nfs_share_path",
+                {
+                    "share_id": data["id"],
+                    "path": path,
+                },
+            )
+        await self.extend(data)
+
+        await self.middleware.call("service.reload", "nfs")
+
+        return data
+
+    @accepts(
+        Int("id"),
+        Patch(
+            "sharingnfs_create",
+            "sharingnfs_update",
+            ("attr", {"update": True})
+        )
+    )
+    async def do_update(self, id, data):
+        verrors = ValidationErrors()
+        old = await self.middleware.call(
+            "datastore.query", self._config.datastore, [("id", "=", id)],
+            {
+                "extend": self._config.datastore_extend,
+                "prefix": self._config.datastore_prefix,
+                "get": True
+            },
+        )
+
+        new = old.copy()
+        new.update(data)
+
+        await self.validate(new, "sharingnfs_update", verrors, old=old)
+
+        if verrors:
+            raise verrors
+
+        await self.compress(new)
+        paths = new.pop("paths")
+        await self.middleware.call(
+            "datastore.update", self._config.datastore, id, new,
+            {
+                "prefix": self._config.datastore_prefix
+            }
+        )
+        await self.middleware.call("datastore.sql", "DELETE FROM sharing_nfs_share_path WHERE share_id = ?", (id,))
```

Comment:
Why SQL and not datastore call?

Reply:
`delete` deletes by prmiary key. Do you want separate call that deletes rows by filter?

Reply:
Well yes, it doesnt fell right to me to use raw SQL while we have an "abstraction" layer, so I am just wondering if there was a reasoning to go that route.